cpp/src/sort.cpp

```diff
@@ -105,6 +331,150 @@ std::unique_ptr<cudf::column> get_split_ind(TaskContext& ctx,
+    exchange_tables.push_back(local_splits_and_metadata);
+  }
+  auto [split_candidates_shared, owners_split] = shuffle(ctx, exchange_tables, nullptr);
+
+  if (local_splits_and_metadata.num_rows() == 0) {
+    // All nodes need to take part in the shuffle (no data here), but the below
+    // cannot search a length 0 table, so return immediately.
+    return nullptr;
+  }
+
+  std::vector<cudf::order> column_orderx(column_order);
+  std::vector<cudf::null_order> null_precedencex(null_precedence);
+  column_orderx.insert(column_orderx.end(), {cudf::order::ASCENDING, cudf::order::ASCENDING});
+  null_precedencex.insert(null_precedencex.end(),
+                          {cudf::null_order::AFTER, cudf::null_order::AFTER});
+
+  // Merge is stable as it includes the rank and index in the keys:
+  // keys(x) to pick columns from splits (which include rank and index):
+  std::vector<cudf::size_type> all_keysx(keys_idx.size() + 2);
+  std::iota(all_keysx.begin(), all_keysx.end(), 0);
+
+  auto split_candidates = cudf::merge(
+    split_candidates_shared, all_keysx, column_orderx, null_precedencex, ctx.stream(), ctx.mr());
+  owners_split.reset();  // No longer need this
+  return std::move(split_candidates);
+}
+
+std::unique_ptr<cudf::table> extract_global_splits(TaskContext& ctx,
+                                                   const cudf::table_view& global_split_candidates)
+{
+  auto split_indices =
+    get_split_ind(ctx, global_split_candidates.num_rows(), ctx.nranks, /* include_start */ false);
+  auto split_value_inds = vector_to_column(split_indices, ctx);
+  auto split_values     = cudf::gather(global_split_candidates,
+                                   split_value_inds->view(),
+                                   cudf::out_of_bounds_policy::DONT_CHECK,
+                                   ctx.stream(),
+                                   ctx.mr());
+  return std::move(split_values);
+}
+
+std::vector<cudf::size_type> find_destination_ranks(
+  TaskContext& ctx,
+  const cudf::table_view& sorted_table,
+  const cudf::table_view& global_split_values,
+  const std::vector<cudf::size_type>& keys_idx,
+  const std::vector<cudf::order>& column_order,
+  const std::vector<cudf::null_order>& null_precedence
+
+)
+{
+  std::vector<cudf::size_type> value_keysx(keys_idx.size() + 1);
+  std::iota(value_keysx.begin(), value_keysx.end(), 0);
+  auto keys_idxx = keys_idx;
+  keys_idxx.push_back(sorted_table.num_columns());
+
+  // Create a column with the same length as sorted table, filled with current rank
+  auto rank_column = create_column<int32_t>(sorted_table.num_rows(), ctx.rank, ctx);
+
+  // Create a new table view by appending the rank column to the sorted table
+  std::vector<cudf::column_view> table_columns;
+  for (int i = 0; i < sorted_table.num_columns(); i++) {
+    table_columns.push_back(sorted_table.column(i));
+  }
+  table_columns.push_back(rank_column->view());
+  auto sorted_table_with_rank = cudf::table_view(table_columns);
```

Comment:
So you append the rank, but not the row-id?  I don't really want to time which strategy is better, so let's keep this and I might open an issue.

FWIW, I suspect we could just use the global row-id at this point if we already add a column.

Either way, I think it's probably better to consider this in a follow up (that is really not urgent).

Reply:
Ah, you are right, we would need that.  I am not sure if you can get it locally via `arr.shape<1>().lo[0]`.  I suspect that works (since we have a concurrent task), but I am not actually fully sure.

Either way, seems like a follow up and you are right, I am not quite sure if it's easy.